third-party/cmake/BuildLibuv.cmake

```diff
@@ -1,18 +1,77 @@
-  DOWNLOAD_COMMAND ${CMAKE_COMMAND}
-    -DPREFIX=${DEPS_BUILD_DIR}
-    -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/libuv
-    -DURL=${LIBUV_URL}
-    -DEXPECTED_SHA256=${LIBUV_SHA256}
-    -DTARGET=libuv
-    -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
-  CONFIGURE_COMMAND sh ${DEPS_BUILD_DIR}/src/libuv/autogen.sh &&
-    ${DEPS_BUILD_DIR}/src/libuv/configure --with-pic --disable-shared
-      --prefix=${DEPS_INSTALL_DIR} --libdir=${DEPS_INSTALL_DIR}/lib
-      CC=${DEPS_C_COMPILER}
-  INSTALL_COMMAND ${MAKE_PRG} install)
+# Reusable function to build libuv, the single argument is
+# the target name. The following variables are used as build
+# commands:
+#   LIBUV_CONFIGURE_COMMAND
+#   LIBUV_BUILD_COMMAND
+#   LIBUV_INSTALL_COMMAND
+#
+function(BuildLibuv TARGET_NAME)
+  ExternalProject_Add(${TARGET_NAME}
+    PREFIX ${DEPS_BUILD_DIR}
+    URL ${LIBUV_URL}
+    DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/libuv
+    DOWNLOAD_COMMAND ${CMAKE_COMMAND}
+      -DPREFIX=${DEPS_BUILD_DIR}
+      -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/libuv
+      -DURL=${LIBUV_URL}
+      -DEXPECTED_SHA256=${LIBUV_SHA256}
+      -DTARGET=${TARGET_NAME}
+      -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
+    CONFIGURE_COMMAND "${LIBUV_CONFIGURE_COMMAND}"
+    BUILD_COMMAND "${LIBUV_BUILD_COMMAND}"
+    INSTALL_COMMAND "${LIBUV_INSTALL_COMMAND}")
+endfunction()
+
+set(UNIX_CFGCMD sh ${DEPS_BUILD_DIR}/src/libuv/autogen.sh &&
+      ${DEPS_BUILD_DIR}/src/libuv/configure --with-pic --disable-shared
+        --prefix=${DEPS_INSTALL_DIR} --libdir=${DEPS_INSTALL_DIR}/lib
+        CC=${DEPS_C_COMPILER}
+        )
+
+if (UNIX)
+  set(LIBUV_CONFIGURE_COMMAND ${UNIX_CFGCMD})
+  set(LIBUV_BUILD_COMMAND "")
+  set(LIBUV_INSTALL_COMMAND ${MAKE_PRG} install)
+  BuildLibUv(libuv) 
+
+elseif (MINGW AND CMAKE_CROSSCOMPILING)
+  # Build libuv for the host
+  set(LIBUV_CONFIGURE_COMMAND sh ${DEPS_BUILD_DIR}/src/libuv_host/autogen.sh && ${DEPS_BUILD_DIR}/src/libuv_host/configure --with-pic --disable-shared --prefix=${HOSTDEPS_INSTALL_DIR} CC=${HOST_C_COMPILER})
+  set(LIBUV_BUILD_COMMAND "")
+  set(LIBUV_INSTALL_COMMAND ${MAKE_PRG} install)
+  BuildLibUv(libuv_host)
+
+  # Build libuv for the target
+  set(LIBUV_CONFIGURE_COMMAND ${UNIX_CFGCMD} --host=${CROSS_TARGET})
+  BuildLibUv(libuv) 
+
+elseif (WIN32 AND MSVC)
+ 
+  find_package(PythonInterp 2.6 REQUIRED)
+  if (NOT PYTHONINTERP_FOUND OR PYTHON_VERSION_MAJOR GREATER 2)
+    message(FATAL_ERROR "Python2 is required to build libuv on windows, use -DPYTHON_EXECUTABLE to set a python interpreter")
+  endif()
+
+  string(FIND ${CMAKE_GENERATOR} Win64 VS_WIN64)
+  if (VS_WIN64 EQUAL -1)
+      set(VS_ARCH x86)
+  else()
+      set(VS_ARCH x64)
+  endif()
+  # FIXME: Remove this when libuv is updated (libuv/libuv#240)
+  set(LIBUV_CONFIGURE_COMMAND git clone https://chromium.googlesource.com/external/gyp.git ${DEPS_BUILD_DIR}/src/libuv/build/gyp)
+  # By default this creates Debug builds
```

Comment:
For MSVC the runtimes between all components must match i.e. I cant build a debug Neovim against a static non-debug libuv - well maybe I can but the compiler will complain. Ideally this should match the build type for the main Neovim (let me have a look at that).